### PR TITLE
Ensure Unique Type Paths for types with generics. 

### PR DESCRIPTION
**Pull request by tadeohepperle**
This PR exposes an `ensure_unique_type_paths` function, like we already use in subxt (in a simpler form). This is intended to be used from subxt, once scale-typegen is a dependency of subxt, to solve [Issue 1247](https://github.com/paritytech/subxt/issues/1247). Please see the last comment of @jsdw on that Issue.

### Testing

Added a test to make sure type paths get amended with numbers. E.g. "Header", "Header2", "Header3", ...

### CODE REVIEW

File: Cargo.lock
Comment (niklasad1) [inline Cargo.lock:1]: you can remove `Cargo.lock` from this repo since it's only used as a library?

File: src/tests/mod.rs [hunk 1]
```diff
@@ -1088,3 +1089,63 @@ fn opt_out_from_default_substitutes() {
...
+            .collect::<Vec<String>>();
+        lines.sort();
+        lines
+    };
+
+    let e1 = sorted_type_paths(&registry);
+    assert_eq!(
+        e1,
+        vec![
+            "",
```
Comment (niklasad1) [inline src/tests/mod.rs:1125]: I don't follow where `""`comes from? 
Is it possible to remove or intended (I don't follow)?

File: src/tests/mod.rs [hunk 2]
```diff
@@ -1088,3 +1089,63 @@ fn opt_out_from_default_substitutes() {
...
+            "",
+            "",
+            "",
+            "scale_typegen::tests::Header",
+            "scale_typegen::tests::Header2",
+            "scale_typegen::tests::Header3",
+        ]
+    );
+    // dbg!(registry);
```
Comment (niklasad1) [inline src/tests/mod.rs:1150]:
```suggestion
```

File: src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
+use std::collections::HashMap;
+
+use scale_info::{form::PortableForm, Field, PortableRegistry, Type, TypeDef, TypeParameter};
+
+/// Checks if the two types are equal, but with a bit of a relaxed condition:
+/// If there is any type id involved that is not equal:
```
Comment (niklasad1) [inline src/utils.rs:6]: Can you rephrase this and the segment below?

I find it a little hard to read.

> Checks if the two types are equal with the relaxed condition that
if the generic type params are equal then the type id's does not need to by equal

**Review by niklasad1 — APPROVED** [review]

File: src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    let mut type_path_lookup = HashMap::<Vec<String>, Type<PortableForm>>::new();
+
+    for ty in &mut types.types {
+        // Ignore types without a path (i.e prelude types).
+        if ty.ty.path.namespace().is_empty() {
+            continue;
+        };
+
+        let mut try_counter = 0;
+        let name_unincremented = ty.ty.path.segments.last().unwrap().clone();
```
Comment (lexnv) [inline src/utils.rs:78]: Would it be possible to have types without path segments?

I would `.expect()` here with a message that tells users metadata is malformed

File: src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            try_counter += 1;
+
+            let name = ty
+                .ty
+                .path
+                .segments
+                .last_mut()
+                .expect("Types have names; qed;");
+
+            *name = format!("{name_unincremented}{}", try_counter + 1);
```
Comment (lexnv) [inline src/utils.rs:99]: This would rename types as: `Name2`, `Name3`.

Would it be better in terms of consistency if we also have a `Name1`? Which should be the first entry from the hash map. 

nit; I would also rewrite a bit the counter to avoid 2 adds, something like:

```rust

let mut name_counter = 1;

if types_equal_extended_to_params(entry, &ty.ty) { .. }

// We know we have at least two types.
if name_counter == 1 {
   hash_entry.name() = ..;
   name_counter += 1;
}
 *name = format!("{name_unincremented}{}", name_counter);
 name_counter += 1;

File: src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        if ty.ty.path.namespace().is_empty() {
+            continue;
+        };
+
+        let mut try_counter = 0;
+        let name_unincremented = ty.ty.path.segments.last().unwrap().clone();
+        loop {
+            let Some(entry) = type_path_lookup.get(&ty.ty.path.segments) else {
+                // Note: clone here not too great, but unavoidable.
+                type_path_lookup.insert(ty.ty.path.segments.clone(), ty.ty.clone());
```
Comment (lexnv) [inline src/utils.rs:82]: This might turn out to be an expensive operation for an edge-case.

Would we be able to avoid cloning here if we change the iteration loop?

We might iterate over the types from the registry by `types.types.iter()` or even by index, instead of doing a `types.types.iter_mut()`. Then store in the hashmap a value of:  `Vec<u32>` instead.
Since this is an edge case, we could iterate a second time over just the values from the hashmap; this time borrowing mut from `types.types.resolve_mut(index)` and modify the names then

File: src/utils.rs (hunk 4 above)
Comment (tadeohepperle) [reply src/utils.rs]: No not really, because those would also have an empty `path.namespace()` and were filtered out before. But I should add a `.expect()` here.

File: src/tests/mod.rs (hunk 1 above)
Comment (tadeohepperle) [reply src/tests/mod.rs]: Oh these lines are for the 4 primitive types registered in the TypeRegistry. It is probably better if I filter them out in this test.

**Review by tadeohepperle — COMMENTED** [review]
Thanks a lot for your reviews guys, I made the function a lot cleaner and more efficient now.

File: Cargo.toml [hunk 7]
```diff
@@ -20,6 +20,7 @@ scale-info = { version = "2.10.0", features = [
     "docs",
     "serde",
 ] }
+smallvec = "1.11.2"
```
Comment (lexnv) [inline Cargo.toml:23]: Nice! 👍

File: src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+/// to generate a single rust type for the two registered typed.
+/// However if they are not equal, we need to deduplicate the type path.
+/// This means modifying the type path of one or both clashing types.
+///
+/// So what types are considered equal?
+/// Types are equal, if their TypeDef is exactly the same.
+/// Types are also considered equal if the TypeDef has the same shape and
+/// all type ids mentioned in the TypeDef are either:
+/// - equal
+/// - or different, but map essentially to the same generic type parameter
```
Comment (lexnv) [inline src/utils.rs:75]: Dumb question: Is the following considered equal? 
```rust
struct First<T, V> {
   first: T,
   second: V,
}

struct Second<T, V> {
   second: T,
   first: V,
}
```

File: src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        .into_iter()
+        .filter_map(|(_, v)| (v.len() > 1).then_some(v));
+
+    // submit to this buffer type ids (u32), where a number (usize) should be added to the type path to distinguish them.
+    let mut renaming_commands: Vec<(u32, usize)> = vec![];
+
+    for type_ids in clashing_type_ids {
+        // Map N types in type_ids to M new type paths, where M <= N.
+        // types with the same structure should map to the same type path after all.
+        let mut types_with_same_structure: Vec<(&Type<PortableForm>, SmallVec<[u32; 2]>)> = vec![];
```
Comment (lexnv) [inline src/utils.rs:30]: We coud have 4 types on the same path, named A, B, C, D.
And let's say that:
- types_equal_extended_to_params(A, C) == true
- types_equal_extended_to_params(B, D) == true
- types_equal_extended_to_params(A, B) == false

Does this still deduplicate/rename all of them?

**Review by lexnv — APPROVED** [review]

File: src/utils.rs (hunk 9 above)
Comment (tadeohepperle) [reply src/utils.rs:30]: I think in this case, we iterate over A, B, C, D:
```
- insert A at index 0,
- insert B at index 1,
- insert C at index 0, because it equals A,
- insert D at index 1, because it equals B but not A.
```
So A and C will get `PathName1` and B and D get `PathName2`. Which should be deduplicated enough.
